iriscasttools/iriscasttools/utils.py

```diff
@@ -0,0 +1,224 @@
+"""
+Provides utility functions to collect energy usage information
+"""
+import subprocess
+import time
+import os
+from pathlib import Path
+import re
+from typing import Dict
+import logging
+import datetime
+
+logger = logging.getLogger(__name__)
+
+
+def retry(
+    retry_on: (Exception,) = None,
+    retries: int = 3,
+    delay: int = 3,
+    backoff: int = 2,
+    retry_logger: logging.Logger = None,
+):
+    """A retry decorator
+
+    Keyword Arguments:
+        retry_on: a set of exceptions that when any occur, will trigger a retry
+        retries: number of retries to perform before returning None
+        delay: number of seconds delay before next retry
+        backoff: a factor to increase delay after every fail
+        retry_logger: logger to log failed attempts
+    """
+
+    def decorator(func):
+        def inner(*args, **kwargs):
+            res = None
+            for i in range(retries + 1):
+                try:
+                    res = func(*args, **kwargs)
+                    break
+                except retry_on:
+                    if i == retries:
+                        if retry_logger:
+                            retry_logger.error(
+                                "function failed and max retries %s exceeded", retries
+                            )
+                        return None
+                seconds = delay + (backoff * i)
+                if retry_logger:
+                    retry_logger.warning(
+                        "function failed to run. Failed attempts: %s. Retrying after %s delay",
+                        i + 1,
+                        seconds,
+                    )
+                time.sleep(seconds)
+            return res
+
+        return inner
+
+    return decorator
+
+
+@retry(retry_on=(AssertionError,), retries=3, delay=3, backoff=2, retry_logger=logger)
+def run_cmd(cmd_args: str):
+    """Run a bash command with given arguments and return output
+
+    Keyword Arguments
+        cmd_args: a string representing bash command to run
+    """
+    with subprocess.Popen(
+        cmd_args, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    ) as out:
+        # assert we did not find any errors
+        assert not out.stderr.read().decode()
```

Comment:
Throw an exception instead of assert

Assert implies a programming error in Python
Exception implies a logical / runtime error, such as the OS not having the package